pkg/kubelet/cm/internal_container_lifecycle_test.go

```diff
@@ -0,0 +1,176 @@
+				memoryManager:   nil,
+				topologyManager: topologymanager.NewFakeManager(),
+			},
+		},
+		{
+			name: "when cpu , memory, topology manager are not nil",
+			i: internalContainerLifecycleImpl{
+				cpuManager:      cpumanager.NewFakeManager(),
+				memoryManager:   memorymanager.NewFakeManager(),
+				topologyManager: topologymanager.NewFakeManager(),
+			},
+		},
+	}
+
+	for _, test := range tests {
+		mockCtrl := gomock.NewController(t)
+
+		pod := &v1.Pod{}
+		container := &v1.Container{}
+		containerId := "test-id"
+		i := test.i
+
+		t.Run(test.name, func(t *testing.T) {
+			mockCPUManager := cpumanager.NewMockManager(mockCtrl)
+			mockMemoryManager := memorymanager.NewMockManager(mockCtrl)
+			mockTopologyManager := topologymanager.NewMockManager(mockCtrl)
+
+			// run the function
+			err := i.PreStartContainer(pod, container, containerId)
+			if err != nil {
+				t.Error("Error Occured!")
+			}
+
+			if i.cpuManager != nil {
+				//i.cpuManager.AddContainer(pod, container, containerId)
+				mockCPUManager.EXPECT().AddContainer(pod, container, containerId).Return(nil).Times(1)
+			} else {
+				mockCPUManager.EXPECT().AddContainer(pod, container, containerId).Return(nil).Times(0)
+			}
+
+			if i.memoryManager != nil {
+				mockMemoryManager.EXPECT().AddContainer(pod, container, containerId).Return(nil).Times(1)
+			} else {
+				mockMemoryManager.EXPECT().AddContainer(pod, container, containerId).Return(nil).Times(0)
+			}
+
+			if i.topologyManager != nil {
+				if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.TopologyManager) {
+					mockTopologyManager.EXPECT().AddContainer(pod, container, containerId).Return(nil).Times(1)
+				} else {
+					mockTopologyManager.EXPECT().AddContainer(pod, container, containerId).Return(nil).Times(0)
+				}
+			} else {
+				mockTopologyManager.EXPECT().AddContainer(pod, container, containerId).Return(nil).Times(0)
+			}
+		})
+	}
+}
+
+func TestPreStopContainer(t *testing.T) {
+
+}
+
+func TestPostStopContainer(t *testing.T) {
+	tests := []struct {
+		name string
+		i    internalContainerLifecycleImpl
+	}{
+		{
+			name: "when topology manager feature gate enabled",
+			i: internalContainerLifecycleImpl{
+				cpuManager:      cpumanager.NewFakeManager(),
+				memoryManager:   nil,
+				topologyManager: topologymanager.NewFakeManager(),
+			},
+		},
+		{
+			name: "when topology manager feature gate disabled",
+			i: internalContainerLifecycleImpl{
+				cpuManager:      nil,
+				memoryManager:   nil,
+				topologyManager: topologymanager.NewFakeManager(),
+			},
+		},
+	}
+
+	mockCtrl := gomock.NewController(t)
+	containerId := "test-id"
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			mockTopologyManager := topologymanager.NewMockManager(mockCtrl)
+			test.i.PostStopContainer(containerId)
+
+			if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.TopologyManager) {
+				mockTopologyManager.EXPECT().RemoveContainer(containerId).Times(1).Return(nil)
```

Comment:
Hello @prakharporwal!
In my opinion, it seems to be an issue because the method expected to be called as below was not called until the end of the test.

```go
mockTopologyManager.EXPECT().RemoveContainer(containerId).Times(1).Return(nil)
```

Furthermore, it appears that the mocked instances such as `mockTopologyManager` are not being used as the test progresses.
Is it okay not to inject the mocked instances into the instances that need to tested like `internalContainerLifecycleImpl`?

Reply:
@21kyu  The function Remove Container is called inside this function	`test.i.PostStopContainer(containerId)`.
Are you saying I should mock the `internalContainerLifecycleImpl` as well ? And then use it  ?

Reply:
Yes, that's right. but as `test.i.PostStopContainer(containerId)` is running, `RemoveContainer` method of `topology manager.NewFakeManager()` is being invoked.
Therefore, I think you should set the `internalContainerLifecycleImpl` as below so that the method of `mockTopologyManager` can be called.

```go
i: internalContainerLifecycleImpl{
 ...
 topologyManager: mockTopologyManager,
}
```